pkg/reconciler/builder/builder.go

```diff
@@ -56,6 +61,7 @@ func NewController(
 		BuildpackLister:        buildpackInformer.Lister(),
 		ClusterBuildpackLister: clusterBuildpackInformer.Lister(),
 		ClusterStackLister:     clusterStackInformer.Lister(),
+		K8sClient:              k8sClient,
```

Comment:
Can we re-org this so that there's a nice separation of concerns where the access level is incrementally stripped down as it passes through abstractions:

1. `cmd/controller/main.go` would construct the `secret.Fetcher{}` struct and pass it to the reconcilers.
    - Only `main` should have raw access to the clients, everything else should be passed in a stripped down interface
1. `pkg/reconciler/builder/builder.go` gets passed a pointer to the `Fetcher` which it then uses to find the secrets for a particular service account. It then passes the list of secrets to the cosign signer
    - Only the reconciler knows what CRs and objects it needs to do a reconcile
1. `pkg/cosign/image_signer.go` gets passed a list of secrets which it then uses to filter out all non-cosign related secrets
    - Only the cosign package knows what's the difference between a regular secret and a cosign secret 

It's the same reason why we pass in all of these `BuildpackLister`, `ClusterBuildpackLister`,  `ClusterStackLister`, etc, instead of just directly calling `c.Client.KpackV1Alpha2().ClusterStack().List()`

Reply:
Could you please elaborate on the pointer to fetcher point? I think I'm not following why it's important to have a pointer there.

Reply:
I've addressed this via 82a7474 and 529ce1f. Looking for some more specific feedback in the creation of the interface for mocking + fake for the `secret.Fetcher`.

Reply:
> Could you please elaborate on the pointer to fetcher point? I think I'm not following why it's important to have a pointer there.

The pointer thing was mainly so that the reconcilers can take an interface instead of a concrete type, which is exactly how you implemented it.

> Looking for some more specific feedback in the creation of the interface for mocking + fake for the `secret.Fetcher`.

I would avoid calling it `IFetcher` and move the interface declaration to the reconcilers. The golang convention is to either export either the interface or the struct, but not both. So if you export the struct, the interface would be defined in the file that uses it, and if you export the interface, the struct would be private.

The convention used in kpack is the former, where we have the Builder and ClusterBuilder reconcilers declare the same `BuilderCreator` interface, and `pkg/cnb/create_builder.go` exports the struct that has the signature.

https://github.com/buildpacks-community/kpack/blob/a8479f9508289d435d513f3681823a096958d357/pkg/reconciler/builder/builder.go#L35-L37
https://github.com/buildpacks-community/kpack/blob/a8479f9508289d435d513f3681823a096958d357/pkg/cnb/create_builder.go#L23-L30

Reply:
I updated this bit, let me know if any further changes are needed!